query-parser/src/parse.rs

```diff
@@ -10,156 +10,191 @@
-    where I: Stream<Item = edn::Value>
-{
-    fn variable() -> FindSpParser<Variable, I> {
-        fn_parser(FindSp::<I>::variable_, "variable")
-    }
-
-    fn variable_(input: I) -> ParseResult<Variable, I> {
-        satisfy_map(|x: edn::Value| Variable::from_value(&x)).parse_stream(input)
-    }
-
-    fn period() -> FindSpParser<(), I> {
-        fn_parser(FindSp::<I>::period_, "period")
-    }
-
-    fn period_(input: I) -> ParseResult<(), I> {
-        satisfy_map(|x: edn::Value| {
-                if let PlainSymbol(ref s) = x {
-                    if s.0.as_str() == "." {
-                        return Some(());
-                    }
-                }
-                return None;
-            })
-            .parse_stream(input)
-    }
-
-    fn ellipsis() -> FindSpParser<(), I> {
-        fn_parser(FindSp::<I>::ellipsis_, "ellipsis")
-    }
-
-    fn ellipsis_(input: I) -> ParseResult<(), I> {
-        satisfy_map(|x: edn::Value| {
-                if let PlainSymbol(ref s) = x {
-                    if s.0.as_str() == "..." {
-                        return Some(());
-                    }
-                }
-                return None;
-            })
-            .parse_stream(input)
-    }
-
-    fn find_scalar() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_scalar_, "find_scalar")
-    }
-
-    fn find_scalar_(input: I) -> ParseResult<FindSpec, I> {
-        (FindSp::variable(), FindSp::period(), eof())
-            .map(|(var, _, _)| FindSpec::FindScalar(Element::Variable(var)))
-            .parse_stream(input)
-    }
-
-    fn find_coll() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_coll_, "find_coll")
-    }
-
-    fn find_coll_(input: I) -> ParseResult<FindSpec, I> {
-        satisfy_unwrap!(edn::Value::Vector, y, {
-                let mut p = (FindSp::variable(), FindSp::ellipsis(), eof())
-                    .map(|(var, _, _)| FindSpec::FindColl(Element::Variable(var)));
-                let r: ParseResult<FindSpec, _> = p.parse_lazy(&y[..]).into();
-                FindSp::to_parsed_value(r)
-            })
-            .parse_stream(input)
-    }
-
-    fn elements() -> FindSpParser<Vec<Element>, I> {
-        fn_parser(FindSp::<I>::elements_, "elements")
-    }
-
-    fn elements_(input: I) -> ParseResult<Vec<Element>, I> {
-        (many1::<Vec<Variable>, _>(FindSp::variable()), eof())
-            .map(|(vars, _)| {
-                vars.into_iter()
-                    .map(Element::Variable)
-                    .collect()
-            })
-            .parse_stream(input)
-    }
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub enum QueryParseError {
+    InvalidInput(edn::Value),
+    EdnParseError(edn::parse::ParseError),
+    MissingField(edn::Keyword),
+    FindParseError(FindParseError),
+}
 
-    fn find_rel() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_rel_, "find_rel")
+impl From<edn::parse::ParseError> for QueryParseError {
```

Comment:
This kind of chaining is where `error-chain` is supposed to shine, but I'm happy with this approach until we have a pattern around `error-chain`.